### ISSUE TITLE
Solve the conundrum of ability points

### ISSUE THREAD

**Issue by pledbrook** [long-war-2/lwotc#23]
As things stand, XCom will likely collect a lot of ability points which could result in an easy route to super soldiers. We should rethink how they work.

For a start, I'm not a big fan of how they are awarded. The current system basically rewards you for certain types of play, but not others. It could potentially also favour some soldier classes over others. Melee fighters are probably hurt the most by this.

I would rather see ability points awards alongside XP. I think we can also open up some interesting player decisions by allowing players to spend ability points to train up soldiers. This would allow them to mitigate the loss of more experienced soldiers. There could be an interesting trade off between building (mildly) super soldiers and being more robust in the face of one-off soldier losses. I don't know whether the ability point system can be repurposed for this.

**Edit** Goals:

 * Turn Ability Points into a consumable/tradable form of XP
 * Mitigate the massive investment in time and XP that constitute high-ranked soldiers that then die
 * Allow for more distinctive soldiers, so a barracks always feels quite different from one campaign to another

Current thoughts/proposal:

 * Remove in-mission AP rewards (the game is just not a good judge of what makes a good play)
 * Award AP at mission end - higher combat intelligence gains more AP
 * Spend AP on non-class skills
 * Add perk rows for: pistols, offensive, defensive
 * Try to align faction soldiers with normal soldiers in promotion mechanics
 * Add one or two staff slots to Training Center that can give promotions at the cost of time and AP, so now you have to decide between super soldiers and a more robust barracks. This can also mitigate late-game squad wipes to some degree.
 * Perhaps use AP for Psi and/or SPARKs

Before going too far down this road, we'll still need to work out what to do with the training center and "AWC" abilities (issue #8).

**Comment by yoshimo** [comment]
rewarding slow strategic play like flanking shots, kill shots from the shadows and high ground feels good. Super soldiers, i am not so sure about. Not every ability can be bougt with the reward points. You still have to choose between different tree options. With the increased  amount of fights maybe increasing the cost a bit is enough. You still need points for more squad members which might already balance tje increased income of ability points

**Comment by pledbrook** [comment]
I don't doubt it feels good (and I assume that was part of the reason for adding the mechanism), but why should clearing cover with a grenade followed by a "flanking" shot get extra benefits over holotargeting an enemy and shooting them through low cover?

The thing is, all of those AP generating scenarios have benefits already, whether through improved aim, improved crit, or whatever. Why add an extra reward that might discourage other valid approaches to defeating the enemy? I'm also a little worried about the potential for farming AP points.

**Comment by yoshimo** [comment]
maybe a few more balanced options for xcom ap are needed. I usually have only a few faction soldiers so i dont see how you would become to powerful with these points. There are just x abilities to buy, so excess points are not helpful

**Comment by GitWadeanthony** [comment]
Would you still use the ability points for the faction soldiers? I personality would like that they level up like the normal LW2 soldiers, give them an extra rank to match and fill out their perk list with the possible abilities they could learn at in 4th row.

Then maybe make the 4th row for all soldiers the only thing you can spend AP on. This way you don't get an Assault with Killer Instinct, Ex Conditioning, Formidable and fortify. In that 4th row at each rank you get one of the [LW2 AWC perks](https://www.ufopaedia.org/index.php/Advanced_Warfare_Center_(LW2)).

The only problem I see with not having the AWC way of training is where to fit the pistol perks but I don't think that's a huge lost. The Templars can get some which were part of their possible abilities added to main tree (Lightning Hands, Quickdraw, and Faceoff) and the rest can be part of the XCOM AWC perks (Gunslinger, Return Fire, Clutch Shot, and Fan Fire).  

Reapers would have 9 free spots to add abilities, Skirmishers 7, and Templars 5.

If you still want to use AP to level up factions soldier however, could we have the LW2 soldier level up that same way?

I know this may be jumping the gun but I had a lot of time thinking about how to add the factions soldier to long war 2.

**Comment by ansabalirai** [comment]
So, here is what I was thinking in terms of options for AP usage:

1) **Reusing LW2 way of getting new abilities**, i.e. training in a tube for a few days and no AP requirement (i.e. just based on rank). This likely means that we either merge infirmary and training center to provide the old AWC functionality, or add two slots in the infirmary or training center and call that good enough.
_Pros_: Less things to balance, Same flexibility in training pistol skills as in LW2, not having to worry about AP farming and creating super soldiers
_Cons_: No real use for all the AP gains that xcom gets except for faction soldiers, implementation issues(?)


2) **Using WOTC way of using AP** to train a few perks from a common pool (probably less than all the ones introduced in LW2). This means that we mostly follow WOTC behavior and no training is required for acquiring additional skills
_Pros_: Simpler utilization of the AP feature, less number of skills available per soldier means chance of getting a supersoldier is small, less implementation work(?)
_Cons_: Supersoldier issue since people might prioritize spending AP on soldiers with better skills, how to train pistol specific skills is still unresolved.


3) **Hybrid option 1**: Have all the LW2 skills for an xcom soldier appear as trainable skills in the training center, i,e. each rank can have 1 offensive, 1 defensive and 1 pistol skill (rather than just 3-5 skills in total as in WOTC). Training is still AP based only (i.e. requires no tube time). Faction soldiers use AP only for skills
_Pros and Cons_:??

4) **Hyrbid option 2**: Same as in option 3, but the cost is a mix of AP and training time. For instance, the AP cost requirement still applies (and can probably be lower than option 2) and the training time can be reduced compared to LW2. The logic is that since you are investing AP resource already, the time cost can be reduced. This also follows the LW2 spriit of people being unavailable when training. _Can also be applicable for faction soldiers?_
_Pros_: Seems more configurable and can be balanced across different difficulties, (_I personally prefer it_ :) )
_Cons_: More complexity than other solutions. Requires more testing to strike the best balance between AP cost and training times

**Comment by yoshimo** [comment]
I like the basic idea of it, but for a better idea how bad the problem is, we need to play the game a bit.
If the gain outweighs the increased amount of soldiers you can spend points on we need to redo it. I lean towards making abilities more expensive and add new ways to get points.
Not sure yet which ability should be trained where.

**Comment by torma** [comment]
I personally lean more and more towards option 1 from @ansabalirai or the proposal from @GitWadeanthony 
In my campaign I'm currently at July 14th with 81 XCOM points which is enough to buy 3-6 extra final tier perks. This doesn't seem like a lot but I also don't have any faction cards increasing ability point gain.    There is a card doubling all gains so I would be sitting at 162.
In addition. All soldiers would be able to have two Master Sergeant level ability if not spending anything while leveling up. This is not including points from XCOM. In my opinion this just doesn't work for LW2.
I would also rather have faction heroes level up the same way as regular soldiers. So choose one ability each tier.
So adding to training slots to the infirmary and use the old LW2 system mostly as is would work the best for me.

**Comment by pledbrook** [comment]
I've updated the description with an initial proposal.

**Comment by gen-martox** [comment]
before you go nuclear on the training center and ability points please lets try a small and simple change first

first add this to

XComGameData.ini
[XComGame.X2StrategyGameRulesetDataStructures]

-AbilityPointCosts[0]=0 ; Squaddie
-AbilityPointCosts[1]=10 ; Lance Corporal
-AbilityPointCosts[2]=11 ; Corporal
-AbilityPointCosts[3]=12 ; Sergeant
-AbilityPointCosts[4]=13 ; Staff Sergeant
-AbilityPointCosts[5]=14 ; Tech Sergeant
-AbilityPointCosts[6]=15 ; Gunnery Sergeant

+AbilityPointCosts[0]=0 ; Squaddie
+AbilityPointCosts[1]=30 ; Lance Corporal
+AbilityPointCosts[2]=31 ; Corporal
+AbilityPointCosts[3]=32 ; Sergeant
+AbilityPointCosts[4]=33 ; Staff Sergeant
+AbilityPointCosts[5]=34 ; Tech Sergeant
+AbilityPointCosts[6]=35 ; Gunnery Sergeant
+AbilityPointCosts[7]=36 ; Master Sergeant

this will change the cost of all perk across the board includeing the faction soldiers so we need to make add this to

XComClassData.ini
[Skirmisher X2SoldierClassTemplate]
-BaseAbilityPointsPerPromotion=10
+BaseAbilityPointsPerPromotion=30

[Reaper X2SoldierClassTemplate]
-BaseAbilityPointsPerPromotion=10
+BaseAbilityPointsPerPromotion=30

[Templar X2SoldierClassTemplate]
-BaseAbilityPointsPerPromotion=10
+BaseAbilityPointsPerPromotion=30

this will keep the faction soldiers as they are now

as it stands all soldier get  3 ability points per promotion and combat intellience increases that

Combat Intelligence, AP gained per rank,  Starting % chance 	
Standard  			 +0   	 	 50% 
Above Average,		+1 			25% 
Gifted 				+2 			15% 
Genius 				+4 			9.5% 
Savant 				+7 			0.5% 


Standard 		        3 + 0 = 3 points per promotion for a tolal of 21 at Master
Above Average 	3 + 1   = 4 points per promotion for a tolal of 28 at Master
Gifted 			3 + 2 = 5 points per promotion for a tolal of 35 at Master
Genius 			3 + 4 = 7 points per promotion for a tolal of 49 at Master
Savant 			3 + 7 = 10 points per promotion for a tolal of 70 at Master

so most soldiers will not be able to more abilities with using the xcom pool and the one they get from combat, which is capped at 2 per mission I believe, those need some changing as well

Genius and Savant levels might need to be pulled back some but they are so rare I dont know how much those would hurt

and making these changes will make Combat Intelligence more important so using the covert actions for increase them more helpful and such those should be changed some as well, increasing how long they take to complete and maybe add some risks would be a good start

I know you want to add the pistol abilities as a fourth row which is good because we could use the New Promotion Screen By Default to decrease there cost back to normal level thus making players having to choice between the expenise "normal" perks and the cheaper pistol perks

you should also see if you could make the officer abilities cost points but I think that would require the Officers Abilities to use the new promotion screen which I dont know how to do sorry

I know this is not a perfect fix but it's better than how it is and better than nuking the training center and sucking all the fun out of it so should at least test this out

**Comment by Kazanir** [comment]
Been thinking about this a lot. Doing promotions via AP sounds both large and hairy and complicated to balanced so I've been thinking about some other ideas along the lines of what martox is saying. Some thoughts about the current negatives and suggestions on possible fixes:

**Negative:** Too easy to build super-soldiers by stacking XCOM AP and buying 3-4 cross-tree abilities on your A-Team.
**Proposed Solutions:**
- Remove XCOM-wide AP. Distribute that AP evenly to soldiers at the end of missions instead.
- Don't remove XCOM-wide AP, but instead restrict it so that it can only be used to cover ~2-4 points of any given ability purchase
- Double the cost of main-class-tree abilities (leaving the randomized XCOM row costing the current amount).

**Negative:** Flexibility of the LW2 AWC in building soldiers (especially the split between offensive/defensive/pistol trees) is lost.
**Proposed Solutions:**
- Add different rows (i.e. like the Offensive/Defensive/Pistol split from LW2 or something else) below the main class tree in the TC
- If combined with the above suggestion, these can all cost "normal" amounts since they are both randomized and not nearly as powerful as combining multiple class skills. This gives you the option of gaining an array of LW2-style utility skills or going the WOTC route and picking up a bonus class skill or two.

**Negative:** Long War 2 balancing required time to train extra abilities on soldiers via the AWC. Allowing all soldiers to get them essentially "for free" as part of the core game is a significant buff to overall XCOM strength.
** Proposed Solutions:**
- Add slots back to the Training Center. This has been suggested a lot and I think it might make sense but I also don't want to have that kill a fun WOTC feature that was available to many more soldiers. Worth thinking about though. I personally always hated the AWC model for various reasons but I might be a rebel on that.

**Comment by micahswab** [comment]
Perhaps individual soldier AP can be used to purchase extra perks immediately, but XCOM AP takes a training slot and additional time investment. This lets experienced soldiers easily get extra perks they've "earned", but curtails dumping shared points into super soldiers.

**Comment by gen-martox** [comment]
I don’t think you can make that work but I don’t know how all the code works for that either. You only get 2 ability points per mission for the xcom pool unless there is a chosen or ruler which gives a few extra on top. I’m not at my pc right now but tomorrow I will post a couple change for the xcom pool which when combined with the changes I proposed will make things better. I also have something else I’m working on but need more time on that before I say more.

**Comment by pledbrook** [comment]
I don't want to sound mean because I do appreciate the ideas and feedback. But I'm not seeing much of the intention behind these proposals. Why make certain changes? What's the aim/end goal?

**Comment by Kazanir** [comment]
My attitude here is sort of, "Balancing is overall very early so let's see what small-to-medium-size tweaks might work to resolve the obvious problem spots so far", without changing the core systems in a big, hairy way. (Not that those ideas aren't also good but sometimes big and far-reaching changes end up being hard to pull off; this is why I took the opposite tack when thinking about solutions.) I think my overall goal would be described as trying to preserve the marquee WOTC systems (if possible) while also making them not obviously imba like the TC currently is.

**Comment by gen-martox** [comment]
That’s the same for me, I want to keep training center the way it is now just in a more balanced way that is easy to change. So that we don’t have a mass army of super soldiers taking the field.

While also making Combat Intelligence more important, the way it currently is I don’t ever by attention to it.

and it doesn’t hurt to take a load you, paledbrook, so you can focus on the more important issues

**Comment by MiserablePileOfBugs** [comment]
I haven't played WOTC a lot so I can't say how broken the actual system is. All I can say is that, from a game design perspective, it looks a bit absurd. I'll try to explain why in the following points:

- If I get this right, the game rewards you with ability points when you accomplish certain actions like taking a flanking shot or killing enemies. So basically, you get bonuses for playing the game well which should be it's own reward because doing so will often prevent you from losing soldiers and help completing your main objective. To me, it looks like a win more type of game design: when you lose you lose but when you win, you win and you win even more (you get the mission reward, loot, experience points AND abilities points). This system has the potential to snowball out of control pretty fast.
- And, like pledbrook mentioned earlier, if this system only encourages certain types of plays, a lot of strategies will become suboptimal and the player might feel frustrated. It's not even an interesting risk vs reward mechanic because flanking the aliens is basically what you are supposed to do.
- Last point: experience and ability points are too similar in their functions. Both are earned while doing the same thing: killing aliens in combat. Both unlock new abilities but the latter kills the interesting choice the player has to make when a soldier levels up: which one of these three abilities should I choose?

TL;DR: interesting choices make for good game design and I don't feel find APs very interesting the way they are at this moment. I'm not suggesting throwing the whole system in the bin but maybe we could change it slightly to only award points at the end of missions (whether you complete the objective or not).

**Comment by gen-martox** [comment]
There is a bit more to getting ability points than that,

1. soldiers get 3 points per promotion depending on combat intelligence, see my earlier post for more on that 

2. you get points for certain actions, flanking shots, ambush kills, height advantage shots, and combo kills. Some will go to the xcom pool and some to the soldier based on combat Intelligence but you only get 2 of these per mission and there is I think a 20-40% chance of getting them

3. You will all so get ability points, which are guaranteed, for kill the chosen, alien rulers, and Julian, and for chasing off a ruler

https://xcom.fandom.com/wiki/Skill_Events

**Comment by pledbrook** [comment]
I'm happy to go with @gen-martox's solution as a quick fix.

However, what I'm interested in with this issue is more what sort of system people would like to eventually end up with and why. What do people like about the current Training Center mechanic and why? What are the gameplay and balance goals?

Also, why do you think I'm planning to "go nuclear" on the training center?

Last point: time is already a valuable resource and very constraining resource. I think removing training times in lieu of using ability points might work fine, as long as ability points have a limited supply and can't be abused. I'm thinking that AP requirements should probably increase on a soldier as more are spent on that soldier for example. That would make it more cost effective to spread the AP around a bit.

**Comment by gen-martox** [comment]
The current system gives players more flexibility with their soldiers abilities, which is good, I think most people tend to use the abilities that will give them the best chance of winning, especially with newer players though I don't know how many "new" players we would get. I think the current system will work fine just need to restrict how many extra perks a soldier can get, which my plan will do. With that said I do think we restrict  certain ability from one another, say DFA and Snap shot, but I don't think they are that many of those, I would point you to Musashi RPD overhaul as he has the code to restrict abilities like that. Also if we can add the pistol abilities as a 4th row, which I'm trying and failing at, and with the AWC abilities, think it will be quite interesting to see where people send their AP at


For most soldiers they will probably get most of their AP from combat and the XCom pool, which we should reduce how many AP can be gained from combat. To that end I have a couple more changes

As I understand it there are 4 tactical event which solder and the xcom pool get AP from but you can only get 2 of these per mission and there is a percent chance of getting these

from GameCore.ini
[XComGame.X2EventListener_AbilityPoints]
MaxTacticalEventAbilityPointsPerBattle=2

[HeightAdvantageShot X2AbilityPointTemplate]
NumPointsAwarded=1
TacticalEvent=true
Chance=20

[FlankingShot X2AbilityPointTemplate]
NumPointsAwarded=1
TacticalEvent=true
Chance=45

[AmbushKill X2AbilityPointTemplate]
NumPointsAwarded=1
TacticalEvent=true
Chance=40

[ComboKill X2AbilityPointTemplate]
NumPointsAwarded=1
TacticalEvent=true
Chance=40

I recommend we start by just reducing the chance for each, making them a less reliable way to get more AP 
I think the height advantage shot will be ok as is but we could reduce it some if need
Flanking shot should be 20-25%
Ambush kill around 15-20%
Combo kill around 15-20%

then there is the non tactical event, meaning killing the chosen, alien rulers, and Julian, which you get on top of the tactical events and are guaranteed to get

from GameCore.ini
[ChosenKilled X2AbilityPointTemplate]
NumPointsAwarded=5
TacticalEvent=false
Chance=100

given the number of missions that one will go on in lwotc, you more likely to see the chosen more often than in WOTC so I think we should either reduce the number of ability points you can get from killing the chosen or we could change the chance of getting them to 50, which I think is better. 

[ChosenDefeated X2AbilityPointTemplate]
NumPointsAwarded=10
TacticalEvent=false
Chance=100

this one is after you permanently kill the chosen, I think it's fine as is. You only get this 3 times and on top of the other changes I suggested it's not even enough for 1 extra ability

this is for rulers
DLC 2 GameCore.ini
[RulerEscapes X2AbilityPointTemplate]
NumPointsAwarded=2
TacticalEvent=false
Chance=100

depending on how lucky/unlucky you are, this one will only come up a couple times a game for I think changing the chance to 75 will be good

[RulerDefeated X2AbilityPointTemplate]
NumPointsAwarded=8
TacticalEvent=false
Chance=100

again this one you can only get 3 times a game but I just wanted to point it out

then there is the 15 AP that you get from killing Julian, which should be fine as is. I mean you only get these if you play that DLC mission and then you have a fair bit of fighting to do to reach Julian, then there is the fact that Julian is a sectopod that you have to fight with laser. so yeah I don't see any reason to change this one


I just had that feeling you were going to nuke the training center after learning a certain player, who shall not be named, was severally abusing the system, tho he is not he first and won't be the last lol

**Comment by GrouchPotato** [comment]
> There is a bit more to getting ability points than that,
> 
> 1. soldiers get 3 points per promotion depending on combat intelligence, see my earlier post for more on that
> 2. you get points for certain actions, flanking shots, ambush kills, height advantage shots, and combo kills. Some will go to the xcom pool and some to the soldier based on combat Intelligence but you only get 2 of these per mission and there is I think a 20-40% chance of getting them
> 3. You will all so get ability points, which are guaranteed, for kill the chosen, alien rulers, and Julian, and for chasing off a ruler
> 
> https://xcom.fandom.com/wiki/Skill_Events

In case this is useful, I've just spent all afternoon searching through the game's config files, source scripts and templates for any indication that AP awards for _combat actions_ are in anyway linked to combat intelligence and I couldn't find anything. I think that wiki page may be out of date.

I could have missed something because there's quite a lot of spaghetti, but I feel certain. From play, I definitely don't remember ever getting anything other than 5 AP for taking down a chosen.

**Comment by pledbrook** [comment]
The system we have is OK for now.
